src/controllers/gasPrice/gasPrice.ts

```diff
@@ -0,0 +1,114 @@
+
+  // network => BundlerGasPrice
+  bundlerGasPrices: { [key: string]: { speeds: GasSpeeds; bundler: BUNDLER } } = {}
+
+  blockGasLimit: bigint | undefined = undefined
+
+  gasPriceTimeout?: { start: any; stop: any }
+
+  constructor(
+    network: Network,
+    provider: RPCProvider,
+    bundlerSwitcher: BundlerSwitcher,
+    getSignAccountOpState: Function
+  ) {
+    super()
+    this.#network = network
+    this.#provider = provider
+    this.#bundlerSwitcher = bundlerSwitcher
+    this.#getSignAccountOpState = getSignAccountOpState
+  }
+
+  async fetch(emitLevelOnFailure: ErrorRef['level'] = 'silent') {
+    const bundler = this.#bundlerSwitcher.getBundler()
+
+    const [gasPriceData, bundlerGas] = await Promise.all([
+      getGasPriceRecommendations(this.#provider, this.#network).catch((e) => {
+        const signAccountOpState = this.#getSignAccountOpState()
+        const estimation = signAccountOpState.estimation as EstimationController
+        const readyToSign = signAccountOpState.readyToSign as boolean
+
+        // Don't display additional errors if the estimation hasn't initially loaded
+        // or there is an estimation error
+        if (!estimation.isLoadingOrFailed() || estimation.estimationRetryError) return null
+
+        const { type } = decodeError(e)
+
+        let message = `We couldn't retrieve the latest network fee information.${
+          // Display this part of the message only if the user can broadcast.
+          readyToSign
+            ? ' If you experience issues broadcasting please select a higher fee speed.'
+            : ''
+        }`
+
+        if (type === ErrorType.ConnectivityError) {
+          message = 'Network connection issue prevented us from retrieving the current network fee.'
+        }
+
+        this.emitError({
+          level: emitLevelOnFailure,
+          message,
+          error: new Error(`Failed to fetch gas price on ${this.#network.name}: ${e?.message}`)
+        })
+        return null
+      }),
+      bundler
+        // no error emits here as most of the time estimation/signing
+        // will work even if this fails
+        .fetchGasPrices(this.#network, () => {})
+        .catch((e) => {
+          this.emitError({
+            level: 'silent',
+            message: "Failed to fetch the bundler's gas price",
+            error: e
+          })
+        })
+    ])
+
+    if (gasPriceData) {
+      if (gasPriceData.gasPrice)
+        this.gasPrices[this.#network.chainId.toString()] = gasPriceData.gasPrice
+      this.blockGasLimit = gasPriceData.blockGasLimit
+    }
+    if (bundlerGas)
+      this.bundlerGasPrices[this.#network.chainId.toString()] = {
+        speeds: bundlerGas,
+        bundler: bundler.getName()
+      }
+
+    this.emitUpdate()
+
+    if (!this.gasPriceTimeout)
+      this.gasPriceTimeout = createRecurringTimeout(() => this.fetch('major'), 12000)
+    this.gasPriceTimeout.stop()
+    this.gasPriceTimeout.start()
```

Comment:
Why do you call stop before start? Also, shouldn't start be called in an else statement?

if (!this.gasPriceTimeout) {
...
} else {
this.gasPriceTimeout.start()
}

Reply:
The easy answer here is that we're guaranteed to not have leeks and double timeouts this way. Performance wise I don't think we're losing something?

Reply:
What I mean is that the start function calls stop() when invoked:
```
  const start = () => {
    if (timeoutId) stop()

    timeoutId = setTimeout(async () => {
      await fn()
      start()
    }, timeout)
  }
```